### DISCUSSION TITLE
How to manually (from CLI) add the first user?

### DISCUSSION THREAD

**Discussion by gyzerok** [wg-easy/wg-easy#660] [Q&A] [answered]
Hello there!

When looking into setting up Wireguard with wg-easy, I got stuck with sort of a chicken and egg problem. Until I have a first peer, I can't connect to the WebUI (because I don't want to expose it to the internet). And until I connect to the WebUI I don't see how to create the first user.

Is there a way perhaps to generate the first peer through the CLI inside the container?

**Comment by eandersons** [comment]
The simplest solution would be to temporarily expose web interface to the Internet, create the first client and then restrict public access to the web interface.

If you are not willing to allow external access to the web interface even temporarily, then you could try to create/modify `wg0.json` to add the first client manually (disclaimer: I have not tried this, so something might not work):

```json
{
  "server": {
    "privateKey": "<server's private key>",
    "publicKey": "<server's public key>",
    "address": "10.0.0.1"
  },
  "clients": {
    "<uuid>": {
    "name": "Client 1",
    "address": "10.0.0.2",
    "privateKey": "<client's private key>",
    "publicKey": "<client's public key>",
    "preSharedKey": "<client's preshared key>",
    "createdAt": "2023-11-22T09:48:07.001Z",
    "updatedAt": "2023-11-22T10:00:08.692Z",
    "enabled": true
  }
}
```

Keys can be generated by following instructions from the [WireGuard official documentation](https://www.wireguard.com/quickstart/#key-generation) (for that WireGuard should be installed on the host or commands can be executed within a WireGuard Easy container):

- to generate a private key:

  ```shell
  $ umask 077
  $ wg genkey > privatekey
  ```
- to generate a public key:
  
  ```shell
  $ wg pubkey < privatekey > publickey
  ```
- to generate a preshared key:

  ```shell
  $ wg genpsk > presharedkey
  ```

To generate an UUID on Linux:

```shell
$ cat /proc/sys/kernel/random/uuid
```

When the necessary changes to `wg0.json` has been made, WireGuard Easy container must be (re)created.

**Reply by eandersons** [reply]
By the way, there is also some kind of API, but it is not meant to use outside of the web interface so it could be tricky to use. You can take a look at #416 and #320 for some insight, ideas and possible solutions..

**Reply by eandersons** [reply, marked as answer]
Another workaround would be to create a SSH tunnel to the web interface (of course, if you have enabled external access to SSH on your server; if not, you could do so temporarily, and that would be safer than temporarily allowing public access to the web interface).

**Comment by gyzerok** [comment]
> The simplest solution would be to temporarily expose web interface to the Internet, create the first client and then restrict public access to the web interface.

Yeah, I've thought of that, but indeed did not want to do this even temporarily.

Thank you for your suggestion! Looks like there is no ready-made way to do what I am looking for. Do you think making a feature request with a friendlier option would make sense? I do love the project otherwise, but think this small issue adds unnecessary friction.

**Reply by eandersons** [reply]
Well, you could create a feature request, but from what I have observed there is a small chance that it will be implemented as:

- the author/maintainer seems to be very busy lately;
- this seems to be some edge case so such functionality would not be widely used.

Even if you would create a corresponding pull request, there is still a small chance that it will be merged.
